# WEEK 3: ASSIGNMENT
Data Analysis and Visulization:
#### What is the Analysis?   A detailed examination of anything complex in order to understand its nature or to determine its essential features

Data Analysis is the process of cleaning, transforming, visualizing, and analyzing the data to gain valuable insights to make more effective business decisions is known as Data Analysis.

## 1-Advanced Pandas
1-https://www.javatpoint.com/pandas-cheat-sheet,
2-https://python.plainenglish.io/a-comprehensive-guide-to-pandas-cheat-sheet-for-data-science-enthusiasts-b6f131ab5284

#### 1.1-Working with Multi-index DataFrames
Multi-index allows you to select more than one row and column in your index.Mean we can combine/add/append multiple arrays,tuples,frames with different values as columns or rows to make one dataset.
We can use various methods of multi-index such as MultiIndex.from_arrays(), MultiIndex.from_tuples(), MultiIndex.from_product(), MultiIndex.from_frame, etc., which helps us to create multiple indexes from arrays, tuples, DataFrame,

##### Importing Necessary libraries

In [1]:
import pandas as pd

#### Creating the Dataframe

In [2]:
# Creating data
Information = {'name': ["Saikat", "Shrestha", "Sandi", "Abinash"],
               
               'Jobs': ["Software Developer", "System Engineer",
                        "Footballer", "Singer"],
               
               'Annual Salary(L.P.A)': [12.4, 5.6, 9.3, 10]}

# Dataframing the whole data
df = pd.DataFrame(Information)

# Showing the above data
df


name                Jobs  Annual Salary(L.P.A)
0    Saikat  Software Developer                  12.4
1  Shrestha     System Engineer                   5.6
2     Sandi          Footballer                   9.3
3   Abinash              Singer                  10.0

#### Performing the multi-indexing

In [3]:
# creating multiple indexes from the dataframe
pd.MultiIndex.from_frame(df)

MultiIndex([(  'Saikat', 'Software Developer', 12.4),
            ('Shrestha',    'System Engineer',  5.6),
            (   'Sandi',         'Footballer',  9.3),
            ( 'Abinash',             'Singer', 10.0)],
           names=['name', 'Jobs', 'Annual Salary(L.P.A)'])

##### here we are appending the columns by creating the dictionary to the dataframe already exists

In [4]:
# making data for dataframing
data = {
    'name': ["Saikat", "Shrestha", "Sandi", "Abinash","Don"],
    'series': ['Peaky blinders', 'Sherlock', 'The crown',
               'Queens Gambit', 'Friends'],
    
    'Ratings': [4.5, 5, 3.9, 4.2, 5],
    
    'Date': [2013, 2010, 2016, 2020, 1994]
}

# Dataframing the whole data created
df2 = pd.DataFrame(data)

# setting first and the second name
# as index column
df2.set_index(["series", "Ratings"], inplace=True,     #as we have 3 things rename,replace,append/add either rows or columns
             append=True, drop=False)
# display the dataframe
print(df2.index)

MultiIndex([(0, 'Peaky blinders', 4.5),
            (1,       'Sherlock', 5.0),
            (2,      'The crown', 3.9),
            (3,  'Queens Gambit', 4.2),
            (4,        'Friends', 5.0)],
           names=[None, 'series', 'Ratings'])


#### 1.2 Merging On 
dataframes based on columns,dataframes with same columns,on multiple columns,on different columns,series into dataframe,multiple dataframes
four types we can do merging: left(take left df and common of right),right(take the right df and common of left),inner(on common column base),outer(take all the columns removing the common columns)

In [5]:
df_merged=df2.merge(df[["name","Jobs","Annual Salary(L.P.A)"]],on="name",how="outer")
df_merged

name          series  Ratings  Date                Jobs  \
0   Abinash   Queens Gambit      4.2  2020              Singer   
1       Don         Friends      5.0  1994                 NaN   
2    Saikat  Peaky blinders      4.5  2013  Software Developer   
3     Sandi       The crown      3.9  2016          Footballer   
4  Shrestha        Sherlock      5.0  2010     System Engineer   

   Annual Salary(L.P.A)  
0                  10.0  
1                   NaN  
2                  12.4  
3                   9.3  
4                   5.6

#### 1.3 joining
difference between the three is that join combines DataFrames based on index while merge combines with both indexes or columns.merge is a useful function for combining Data Frames on specific columns. Join is a useful function for combining along indexes or key columns. Lastly, concat is most useful for combining along either axis or for multi-indexing.

In [6]:
df_joined=df.set_index('name').join(df2.set_index('name'), how='outer')

df_joined

Jobs  Annual Salary(L.P.A)          series  Ratings  \
name                                                                          
Abinash               Singer                  10.0   Queens Gambit      4.2   
Don                      NaN                   NaN         Friends      5.0   
Saikat    Software Developer                  12.4  Peaky blinders      4.5   
Sandi             Footballer                   9.3       The crown      3.9   
Shrestha     System Engineer                   5.6        Sherlock      5.0   

          Date  
name            
Abinash   2020  
Don       1994  
Saikat    2013  
Sandi     2016  
Shrestha  2010

#### 1.4 Concatenation on datasets
combining the dataframes

In [7]:
df_concatenated=pd.concat([df,df2],axis=1)
df_concatenated

name                Jobs  Annual Salary(L.P.A)  \
0                           Saikat  Software Developer                  12.4   
1                         Shrestha     System Engineer                   5.6   
2                            Sandi          Footballer                   9.3   
3                          Abinash              Singer                  10.0   
(0, Peaky blinders, 4.5)       NaN                 NaN                   NaN   
(1, Sherlock, 5.0)             NaN                 NaN                   NaN   
(2, The crown, 3.9)            NaN                 NaN                   NaN   
(3, Queens Gambit, 4.2)        NaN                 NaN                   NaN   
(4, Friends, 5.0)              NaN                 NaN                   NaN   

                              name          series  Ratings    Date  
0                              NaN             NaN      NaN     NaN  
1                              NaN             NaN      NaN     NaN  
2                              NaN             NaN      NaN     NaN  
3                              NaN             NaN      NaN     NaN  
(0, Peaky blinders, 4.5)    Saikat  Peaky blinders      4.5  2013.0  
(1, Sherlock, 5.0)        Shrestha        Sherlock      5.0  2010.0  
(2, The crown, 3.9)          Sandi       The crown      3.9  2016.0  
(3, Queens Gambit, 4.2)    Abinash   Queens Gambit      4.2  2020.0  
(4, Friends, 5.0)              Don         Friends      5.0  1994.0

##### Importing data file

In [ ]:
df=pd.read_csv("E:\\Programming\\internship\\week3assignment\\merch_sales.csv")

#### 1.5 Handling missing data and cleaning up data

In [9]:
import numpy as np

# Sample DataFrame with missing values
df = pd.DataFrame({
    'name': ['Saikat', 'Shrestha', np.nan, 'Abinash', 'Don'],
    'Jobs': ['Software Developer', 'System Engineer', 'Footballer', np.nan, 'Singer'],
    'Annual Salary(L.P.A)': [12.4, 5.6, 9.3, np.nan, 10]
})

# 1. Identify missing data
print("Missing data count per column:")
print(df.isnull().sum())

# 2. Fill missing values
df['name'] = df['name'].fillna('Unknown')  # Fill 'name' with 'Unknown'
df['Jobs'] = df['Jobs'].fillna('Not Specified')  # Fill 'Jobs' with 'Not Specified'
df['Annual Salary(L.P.A)'] = df['Annual Salary(L.P.A)'].fillna(df['Annual Salary(L.P.A)'].mean())  # Fill salary with mean

# 3. Remove duplicates
df = df.drop_duplicates()

# 4. Handle outliers (capping in this case)
df['Annual Salary(L.P.A)'] = df['Annual Salary(L.P.A)'].clip(lower=5, upper=15)

# 5. Rename columns for clarity
df.rename(columns={'Annual Salary(L.P.A)': 'Salary'}, inplace=True)

# Display cleaned dataframe
print("\nCleaned DataFrame:")
print(df)


Missing data count per column:
name                    1
Jobs                    1
Annual Salary(L.P.A)    1
dtype: int64

Cleaned DataFrame:
       name                Jobs  Salary
0    Saikat  Software Developer  12.400
1  Shrestha     System Engineer   5.600
2   Unknown          Footballer   9.300
3   Abinash       Not Specified   9.325
4       Don              Singer  10.000


#### 1.6- Grouping and Aggregate using custom functions

In [10]:

# Sample DataFrame
data = {
    'name': ['Saikat', 'Shrestha', 'Sandi', 'Abinash', 'Don'],
    'Jobs': ['Software Developer', 'System Engineer', 'Footballer', 'Singer', 'Footballer'],
    'Annual Salary(L.P.A)': [12.4, 5.6, 9.3, 10.0, 7.5],
    'Age': [28, 30, 25, 22, 27]
}

df = pd.DataFrame(data)
print(df)


       name                Jobs  Annual Salary(L.P.A)  Age
0    Saikat  Software Developer                  12.4   28
1  Shrestha     System Engineer                   5.6   30
2     Sandi          Footballer                   9.3   25
3   Abinash              Singer                  10.0   22
4       Don          Footballer                   7.5   27


In [14]:
grouped=df.groupby('Jobs')
grouped

In [12]:
aggregated=grouped['Annual Salary(L.P.A)'].mean() #similar for sum,min,max
aggregated

Jobs
Footballer             8.4
Singer                10.0
Software Developer    12.4
System Engineer        5.6
Name: Annual Salary(L.P.A), dtype: float64

In [15]:
aggregated=grouped['Annual Salary(L.P.A)'].agg(['min','max','sum'])
aggregated

min   max   sum
Jobs                                
Footballer           7.5   9.3  16.8
Singer              10.0  10.0  10.0
Software Developer  12.4  12.4  12.4
System Engineer      5.6   5.6   5.6

## 2- Numpy Applications

#### Eigenvalues & Eigenvector
**Scalar:** 0-dimensional, just a single number.Example: 5, 3.14, -42<br>
**Vector:** 1-dimensional, an array of numbers. Example: [1, 2, 3], [0, 3.5, 6, 9]<br>
**Matrix:** 2-dimensional, an array of numbers arranged in rows and columns.Example: [[1, 2], [3, 4]], [[5, 6, 7], [8, 9, 10]] <br>
Here basically we have matrixlets say A=[1 2 3] and vector which is also a matrix X=[4 5 6 ],& when we compute them by performing the dot product ,at the end we have some value which is either 0 or some number that's what we can say our   
Compute the product AX
 for
AX=kX
  where  k
  is some scalar,When this equation holds for some  X
  and  k
 , we call the scalar  k
  an eigenvalue of  A
 . We often use the special symbol  λ
  instead of  k
  when referring to eigenvalues. In Example  7.1.1
 , the values  10
  and  0
  are eigenvalues for the matrix  A
  and we can label these as  λ1=10
  and  λ2=0
 
When  AX=λX
  for some  X≠0
 , we call such an  X
  an eigenvector of the matrix  A
 . The eigenvectors of  A
  are associated to an eigenvalue. Hence, if  λ1
  is an eigenvalue of  A
  and  AX=λ1X
 , we can label this eigenvector as  X1
 . Note again that in order to be an eigenvector,  X
  must be nonzero.
Let’s say you have a matrix 
𝐴
A representing some transformation in space. If you multiply this matrix by a vector 
𝑋
X, it will generally rotate and scale 
𝑋
X. But if 
𝑋
X is an eigenvector of 
𝐴
A, then multiplying 
𝐴
A by 
𝑋
X will only scale 
𝑋
X by a factor 
𝜆
λ, without changing its direction. In other words:

𝐴
𝑋
=
𝜆
𝑋
AX=λX
So, the eigenvalue 
𝜆
λ represents how much the vector is stretched or compressed, and the eigenvector 
𝑋
X represents the direction that stays unchanged under the transformation by matrix 
𝐴
This equation means that when we apply the matrix 
𝐴
A to the vector 
𝑋
X, the vector 
𝑋
X only gets scaled by 
𝜆
λ (but not rotated or changed direction). The eigenvector 
𝑋
X remains in the same direction after being transformed by 
𝐴
A, but its length is scaled by the corresponding eigenvalue 
𝜆
λ.
  for furthere understanding you can look at here this link    https://byjus.com/maths/eigen-values/
  ##### Why/Purpose of eigenvalues and vector
**1. Simplifying Matrix Operations:**
Eigenvalues and eigenvectors allow us to diagonalize a matrix (if possible), which simplifies matrix operations. When a matrix is diagonalized, it is represented as a product of its eigenvectors and eigenvalues, making many matrix calculations, like exponentiation, much easier.

Example: In physics or in systems analysis, diagonalization helps reduce complex systems of linear equations to simpler forms.

**2. Data Reduction (PCA in Machine Learning):**
Eigenvalues and eigenvectors are used in Principal Component Analysis (PCA), a common technique in machine learning and data science for dimensionality reduction. PCA helps identify the "principal components" of the data, which are the directions (eigenvectors) that explain the most variance in the data. The eigenvalues tell you how much variance each principal component explains.

***Why it matters:*** In large datasets, many features (dimensions) might be redundant. PCA reduces the dimensions of the data while retaining most of the information, which makes the data easier to work with (e.g., faster training in machine learning).

In [16]:
import numpy as np

# Define a square matrix
A = np.array([[4, -2], 
              [1,  1]])

# Compute the eigenvalues and eigenvectors
eigenvalues, eigenvectors = np.linalg.eig(A)

print("Eigenvalues:", eigenvalues)
print("Eigenvectors:\n", eigenvectors)


Eigenvalues: [3. 2.]
Eigenvectors:
 [[0.89442719 0.70710678]
 [0.4472136  0.70710678]]


#### Singular Value Decomposition (SVD)
The Singular Value Decomposition of a matrix is a factorization of the matrix into three matrices. Thus, the singular value decomposition of matrix A can be expressed in terms of the factorization of A into the product of three matrices as A = UDVT

In [39]:
# Define a matrix
A = np.array([[1, 2], 
              [3, 4]])

# Perform Singular Value Decomposition
U, S, Vt = np.linalg.svd(A)

print("U:\n", U)
print("S (singular values):\n", S)
print("Vt (transposed V):\n", Vt)


U:
 [[-0.40455358 -0.9145143 ]
 [-0.9145143   0.40455358]]
S (singular values):
 [5.4649857  0.36596619]
Vt (transposed V):
 [[-0.57604844 -0.81741556]
 [ 0.81741556 -0.57604844]]


#### Vectorized Operations
NumPy allows you to perform element-wise operations on arrays or matrices, making computations faster and more efficient than traditional Python loops. This is called vectorization.

In [ ]:
# Define two matrices
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])

# Perform element-wise addition
C = A + B
print("Element-wise addition:\n", C)

# Perform element-wise multiplication
D = A * B
print("Element-wise multiplication:\n", D)

# Perform matrix multiplication (dot product)
E = np.dot(A, B)
print("Matrix multiplication (dot product):\n", E)


#### Performance Optimization (Avoid Loops with NumPy)
One of the main advantages of NumPy is performance. NumPy operations are optimized and written in C, which makes them much faster than traditional Python loops.

In [ ]:
# Define a large matrix
A = np.random.rand(1000, 1000)

# Using a loop to square every element (slow)
result_loop = np.zeros_like(A)
for i in range(A.shape[0]):
    for j in range(A.shape[1]):
        result_loop[i, j] = A[i, j] ** 2

# Using a vectorized operation to square every element (fast)
result_vectorized = A ** 2

print("Result (loop) and (vectorized) are the same:", np.allclose(result_loop, result_vectorized))


#### Broadcasting
Broadcasting is a powerful feature in NumPy that allows you to perform arithmetic operations on arrays of different shapes. NumPy automatically expands the smaller array to match the shape of the larger one in a process known as broadcasting.

In [17]:
# Define a 2D matrix and a 1D vector
A = np.array([[1, 2], [3, 4]])
B = np.array([1, 2])

# Broadcasting allows adding a 1D vector to a 2D matrix
C = A + B
print("Broadcasting result:\n", C)


Broadcasting result:
 [[2 4]
 [4 6]]


#### Reshaping
Reshaping allows you to change the shape of a NumPy array without changing its data. This is useful when you need to prepare data for matrix operations or machine learning algorithms.

In [ ]:
# Define a 1D array
A = np.array([1, 2, 3, 4, 5, 6])

# Reshape it into a 2x3 matrix
B = A.reshape(2, 3)
print("Reshaped matrix (2x3):\n", B)

# Flatten the matrix back into a 1D array
C = B.flatten()
print("Flattened array:\n", C)


#### Advanced Operations (Determinants, Inverses, etc.)
You can also perform other advanced matrix operations such as computing the determinant or the inverse of a matrix.

##### Determinant
its basically done to check the matrix either it is feasible or not for furthere operations like inversing etc.
its a scalar value after determinant it means it can be used otherwise not.

In [19]:
# Define a square matrix
A = np.array([[4, 2], [3, 1]])

# Calculate the determinant
det = np.linalg.det(A)
print("Determinant of A:", det)


Determinant of A: -2.0


##### Inverse
Since we can’t do division with matrix so we use inverse of a matrix for division.

For example: If you have to divide 10 pens to 2 people you can simply do 10/2=5 or you can find the inverse of 2 i.e 0.5 and multiply it with 10 which gives (10*0.5 =5)
Similarly if we have to calculate value of matrix X in the equation X A= B

Then, X=B/A (but since we can’t do division in matrix we can multiply B with inverse of A).

So this can be written as X A (A)^-1 = B (A)^-1 (multiplying both sides with the inverse of A).

The above will give X*I = B(A)^-1 (I=Identity matrix,we can replace it with it’s magnitude which is 1 as A(A)^-1 = I).
As 2.1/2 =1 similarly A.A-1=1,

In [18]:
# Compute the inverse of a matrix
A_inv = np.linalg.inv(A)
print("Inverse of A:\n", A_inv)


Inverse of A:
 [[-2.   1. ]
 [ 1.5 -0.5]]


## 3- Data Visualization with matplotlib and seaborn
**Seaborn vs Matplotlib**
**Matplotlib** is a low-level plotting library that provides a high degree of control over individual elements. Even for basic functionalities, it requires more code.

Whereas **seaborn** is a high level library for visualization and requires less coding compared to matplotli.

**Matplotlib** lets users customize the appearances of plots, including color and styles.

**Seaborn** has in-built themes and color palettes making it easier for users to create visually appealing plots.

**Matplotlib** can work with pandas but users may need to manipulate data for certain type of plots.

**Seaborn** is very much flexible with pandas and it doesn’t require as much manipulation as matplotlib.

**Annotations** allow users to label data points and indicate trends or add descriptions to different parts of a plot.

**Both seaborn and matplotlib have style themes.**

There are **five themes** in seaborn:
`white,
dark,
whitegrid,
darkgrid,
ticks`
We will use style theme from matplotlib. Matplotlib gives 26 styles which you can be seen with the `plt.style.available` method.
**Output:**
`['Solarize_Light2',
 '_classic_test_patch',
 'bmh',
 'classic',
 'dark_background',
 'fast',
 'fivethirtyeight',
 'ggplot',
 'grayscale',
 'seaborn',
 'seaborn-bright',
 'seaborn-colorblind',
 'seaborn-dark',
 'seaborn-dark-palette',
 'seaborn-darkgrid',
 'seaborn-deep',
 'seaborn-muted',
 'seaborn-notebook',
 'seaborn-paper',
 'seaborn-pastel',
 'seaborn-poster',
 'seaborn-talk',
 'seaborn-ticks',
 'seaborn-white',
 'seaborn-whitegrid',
 'tableau-colorblind10']`

### What is Visualization?
Visualization refers to techniques used to communicate both abstract and concrete,behaviour,information, ideas by creating images, diagrams, or animations of objects

### 3.1- Data Visualization Explanation
Data visualization uses graphical representations like graphs, charts, and maps to simplify complex data, making it easier to understand and analyze. Visuals allow data scientists to summarize thousands of rows and columns of complex data and put it in an understandable and accessible format. It helps identify patterns, correlations, and outliers, providing a more effective analysis than tables or descriptive statistics. Data visualization is crucial in decision-making, enabling data analysts, scientists, and engineers to communicate insights to non-technical stakeholders and inform actions, such as A/B testing or addressing bias in models like ChatGPT.  
Link to understand more about the data visualization: [What is Data Visualization?](https://www.couchbase.com/blog/what-is-data-analysis/)

#### A comprehensive breakdown of Data Analysis and Data Visualization concepts, organized into a structured tree format to clarify the relationship between different elements:

#### 1. Data Types
- **Quantitative Data**: Numerical values, used for statistical analysis.
- **Qualitative Data**: Descriptive data, typically non-numerical and used for interpretive analysis.

#### 2. Types of Data Analysis
- **Univariate Analysis**: Univariate Analysis for Continuous Variables and Categorical Variables.
- **Bivariate Analysis**: Bivariate Analysis for Continuous Variable vs Continuous Variable, Categorical Variable vs Categorical Variable.
- **Multivariate Analysis**: Multivariate Analysis for Numerical-Numerical,Numerical-Categorical Variables.

#### 3. Data Analysis Techniques
- **Descriptive Analysis**: Summarizes main features (e.g., mean, median, mode).  
  **Tools**: Histograms, Bar Charts.  
- **Diagnostic Analysis**: Investigates causes and patterns (e.g., regression, correlation).  
  **Tools**: Scatter plots, Heatmaps.  
- **Predictive Analysis**: Forecasts future outcomes using historical data.  
  **Tools**: Line plots, Forecasting models.  
- **Prescriptive Analysis**: Provides recommendations based on predictive insights.  
  **Tools**: Optimization models, Decision trees.  

#### 4. Types of Data Analysis Methods
- **Statistical Analysis**: Applies statistical techniques (e.g., regression, hypothesis testing).  
  **Tools**: R, Python (scikit-learn, NumPy).  
- **Data Mining**: Discovers patterns from large datasets.  
  **Tools**: Clustering, Classification algorithms.  
- **Text Mining**: Analyzes unstructured textual data.  
  **Tools**: Natural Language Processing (NLP), sentiment analysis.  
- **Time Series Analysis**: Focuses on data points collected over time.  
  **Tools**: Forecasting, Decomposition.

#### 5. Key Data Visualization Techniques
- **Distribution**: Analyzing how data points are spread.  
  **Visualizations**: Histograms, Boxplots.  
- **Correlation**: Identifying relationships between variables.  
  **Visualizations**: Scatter plots, Heatmaps.  
- **Ranking**: Comparing quantities in order.  
  **Visualizations**: Bar charts, Dot plots.  
- **Part-of-Whole**: Showing proportions of a total.  
  **Visualizations**: Pie charts, Stacked Bar charts.  
- **Evolution**: Analyzing changes over time.  
  **Visualizations**: Line charts, Area charts.  
- **Map**: Geospatial data visualization.  
  **Visualizations**: Choropleth maps, Heat maps.  
- **Networks**: Analyzing connections or relationships.  
  **Visualizations**: Network graphs, Force-directed graphs.

#### 6. Deciding the Right Technique
**Question to Ask**:  
- How many variables are you analyzing?  
  - **Univariate Analysis**: Analyzing a single variable to understand its distribution, central tendency, and variability.  
  - **Bivariate Analysis**: Analyzing the relationship between two variables, often looking for correlations or associations.  
  - **Multivariate Analysis**: Analyzing the relationships between three or more variables simultaneously.

**What do you want to analyze?**  
- Use suitable techniques for distribution, correlation, ranking, etc.  
- Choose appropriate visualizations depending on analysis goals.

#### 7. Data Analysis Process
- **Define Objective**: Understand what you want to achieve.
- **Prepare and Explore Data**: Clean data, check for outliers.
- **Apply Analysis Techniques**: Choose methods (statistical, machine learning, etc.).
- **Interpret Results**: Derive meaningful insights.
- **Communicate Findings**: Use visualizations and reports to convey results effectively.

#### 8. Common Data Analysis Tools
- **Spreadsheet Software**: Basic analysis (Excel, Google Sheets).
- **Business Intelligence Platforms**: Interactive dashboards (Power BI, Tableau).
- **Programming Languages**: Advanced analysis (Python, R).
- **Cloud-Based Platforms**: Scalable environments for large datasets (Google Cloud, AWS).
- **Text Analytics Tools**: Analyzing unstructured data (NLP libraries, sentiment analysis tools).

#### 9. Importance of Data Analysis
- **Decision Making**: Provides insights for informed choices.
- **Problem Solving**: Identifies root causes and optimizes processes.
- **Performance Evaluation**: Measures success and evaluates KPIs.
- **Risk Management**: Identifies and mitigates potential risks.
- **Gathering Insights**: Drives innovation and strategic actions.

This tree structure organizes data analysis and visualization concepts systematically, making it easier to understand how to approach data analysis, choose the right techniques, and select appropriate visualizations based on your data and goals.


### 3.2- Structured Flow for Implementing Data Analysis

The process of implementing data analysis can be broken down into clear steps, starting from problem identification and leading up to decision-making based on the analysis. Below is a structured approach for implementing data analysis:

#### Step 1: Define the Objective
- **Action**: Clearly identify the question or problem you want to solve with data.
- **Example**: "What are the key factors influencing customer churn?"

#### Step 2: Collect and Prepare the Data
- **Action**: Gather data relevant to the analysis.  
  Ensure data quality (handle missing values, duplicates, or irrelevant information).
- **Example**: Collect customer demographic data, transaction history, and support interaction logs.

#### Step 3: Identify the Type of Data
- **Action**: Classify your data into Quantitative or Qualitative:
  - **Quantitative**: Numerical, measurable (e.g., age, sales, etc.).
  - **Qualitative**: Descriptive, non-numerical (e.g., customer feedback, reviews).
- **Example**: Customer age (quantitative), customer feedback (qualitative).

#### Step 4: Choose the Type of Data Analysis
- **Action**: Select the type of analysis based on your objective:
  - **Univariate Analysis** (for one variable): Summarize basic statistics (mean, median, etc.).
  - **Bivariate Analysis** (for two variables): Identify relationships (e.g., correlation).
  - **Multivariate Analysis** (for multiple variables): Explore interactions among many variables.
- **Example**: Analyze customer churn using Bivariate analysis (e.g., age vs. churn, tenure vs. churn).

#### Step 5: Select the Appropriate Data Analysis Technique
- **Action**: Based on the data type and analysis, choose the most suitable technique:
  - **Descriptive**: Summarize key statistics.
  - **Diagnostic**: Identify reasons or relationships.
  - **Predictive**: Forecast future behavior or outcomes.
  - **Prescriptive**: Provide recommendations for optimal action.
- **Example**: For customer churn prediction, Predictive analysis using a regression model or machine learning.

#### Step 6: Visualize the Data
- **Action**: Choose appropriate visualizations based on your analysis type:
  - **Distribution**: Histograms, Boxplots.
  - **Correlation**: Scatter Plots, Heatmaps.
  - **Ranking**: Bar Charts, Dot Plots.
  - **Part-of-Whole**: Pie Charts, Stacked Bars.
  - **Evolution**: Line Charts, Area Charts.
  - **Map**: Geospatial Visualization.
  - **Networks**: Network Graphs.
- **Example**: Use scatter plots to visualize the relationship between age and churn rate, or line charts to track churn over time.

#### Step 7: Apply the Chosen Analysis Method
- **Action**: Apply the selected technique using appropriate tools:
  - **Statistical Methods** (e.g., hypothesis testing, regression).
  - **Data Mining** (e.g., clustering, classification).
  - **Text Mining** (e.g., sentiment analysis).
  - **Time Series** (e.g., forecasting, decomposition).
- **Example**: Apply predictive analysis using regression or machine learning algorithms to predict churn.

#### Step 8: Interpret the Results
- **Action**: Review the output of your analysis and interpret the results:
  - What patterns, trends, or correlations have emerged?
  - What are the key insights, and what do they mean in the context of your problem?
- **Example**: Identify that age and customer tenure are strong predictors of churn, with younger customers and those with shorter tenure more likely to leave.

#### Step 9: Make Data-Driven Decisions
- **Action**: Use the insights to make informed decisions or recommendations:
  - Suggest improvements, strategies, or actions based on the analysis.
- **Example**: Develop a retention strategy focusing on younger customers and offering incentives to long-term customers to reduce churn.

#### Step 10: Communicate the Findings
- **Action**: Present your findings clearly to stakeholders:
  - Use visualizations to highlight key points (charts, graphs).
  - Provide a clear narrative of the insights and actionable recommendations.
- **Example**: Present a report and dashboard to executives, showcasing the key churn drivers, visualizing trends, and suggesting actions for retention.

#### Step 11: Take Action & Monitor
- **Action**: Implement the suggested actions and monitor their effectiveness over time:
  - Use **Prescriptive Analysis** to optimize strategies and actions.
  - Monitor outcomes to see if the decisions lead to improved results.
- **Example**: After implementing retention strategies, monitor churn rates and evaluate the success of your interventions.

#### Step 12: Iteration and Refinement
- **Action**: Data analysis is an iterative process:
  - As new data comes in, revisit the analysis to refine insights or improve predictions.
- **Example**: Track churn metrics monthly, refining your predictive model with new customer data.

#### Tools & Resources Needed:
- **Basic Tools**: Excel, Google Sheets (for simple analysis).
- **Advanced Tools**: R, Python (for machine learning and deep analysis).
- **Visualization Tools**: Tableau, Power BI (for interactive reports and dashboards).
- **Cloud Analytics**: AWS, Azure (for large-scale data processing).

### Summary of the Flow:
- **Define the objective → Collect and prepare data → Identify data types → Choose analysis type → Select analysis technique → Visualize the data → Apply analysis methods → Interpret the results → Make decisions → Communicate findings → Take action → Refine and iterate**.

This flow ensures a systematic approach to data analysis and visualization, enabling you to extract meaningful insights, make informed decisions, and optimize outcomes for your organization or research.


### 3.3- Plots explanation and their use case
`Line Plot: Use plt.plot() to visualize continuous data.
Scatter Plot: Use plt.scatter() to show relationships between two variables.
Bar Plot: Use plt.bar() for categorical data.
Histogram: Use plt.hist() to show the distribution of data.
Heatmap: Use sns.heatmap() for 2D data with color coding.
Pair Plot: Use sns.pairplot() for visualizing pairwise relationships.
Joint Plot: Use sns.jointplot() for examining the relationship between two variables.
Customizing Plots: Add annotations using plt.annotate(), change styles using plt.style.use(), and apply themes using sns.set_theme().
Saving Plots: Use plt.savefig() to save plots in different formats like .png, .svg.`

#### libraries to Import

In [48]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd


#### creating the sample data

In [54]:
# Set a theme for Seaborn
sns.set_theme(style="whitegrid")

# Sample Data
x = np.linspace(0, 10, 100)
y = np.sin(x)
categories = ['A', 'B', 'C', 'D']
values = [5, 7, 3, 8]
data = np.random.randn(1000)
df = sns.load_dataset('iris')

#### Line Plot
A line plot is used to display data points in a continuous line, typically for time-series or continuous data.

In [ ]:
# 1. Line Plot with Customization (Annotations, Styles, and Themes)
plt.figure(figsize=(8, 6))
plt.plot(x, y, color='b', label='sin(x)')
plt.title('Line Plot with Customizations')
plt.xlabel('X axis')
plt.ylabel('Y axis')
plt.grid(True)
plt.legend()

# Add annotation
plt.annotate('Peak', xy=(np.pi/2, 1), xytext=(3, 0.5),
             arrowprops=dict(facecolor='red', arrowstyle='->'))
plt.show()

#### Scatter Plot
A scatter plot is used to show the relationship between two variables, where each point represents one observation.

In [ ]:
# 2. Scatter Plot with Customization (Annotations, Styles, and Themes)
plt.figure(figsize=(8, 6))
plt.scatter(x[:50], y[:50], color='green', label='Data points', alpha=0.6)
plt.title('Scatter Plot with Customizations')
plt.xlabel('X axis')
plt.ylabel('Y axis')
plt.legend()
plt.show()

#### Bar Plot
A bar plot is used to represent categorical data with rectangular bars where the length of each bar represents a category's value.

In [ ]:
# 3. Bar Plot with Customization (Annotations, Styles, and Themes)
plt.figure(figsize=(8, 6))
plt.bar(categories, values, color='orange')
plt.title('Bar Plot with Customizations')
plt.xlabel('Categories')
plt.ylabel('Values')

# Add annotation
for i, v in enumerate(values):
    plt.text(i, v + 0.2, str(v), ha='center', color='black')
plt.show()

#### Histogram
A histogram is used to represent the distribution of a dataset by dividing the data into intervals or bins.

In [ ]:
# 4. Histogram with Customization (Annotations, Styles, and Themes)
plt.figure(figsize=(8, 6))
plt.hist(data, bins=30, color='purple', edgecolor='black')
plt.title('Histogram with Customizations')
plt.xlabel('Value')
plt.ylabel('Frequency')
plt.show()

#### Heatmap
A heatmap is a 2D representation of data where individual values are represented by colors.

In [ ]:
# 5. Heatmap with Customization (Annotations, Styles, and Themes)
plt.figure(figsize=(8, 6))
heatmap_data = np.random.rand(10, 12)
sns.heatmap(heatmap_data, annot=True, cmap='coolwarm', cbar=True)
plt.title('Heatmap with Customizations')
plt.show()

#### Pair Plot
A pair plot shows pairwise relationships in a dataset. It plots each variable against every other variable in the dataset.

In [ ]:
# 6. Pair Plot with Customization (Annotations, Styles, and Themes)
sns.pairplot(df, hue='species', palette='Set2', markers='o')
plt.title('Pair Plot with Customizations')
plt.show()

#### Joint Plot
A joint plot is used to show the relationship between two variables along with their marginal distributions.

In [ ]:
# Save the Joint Plot as a PNG file
plt.figure(figsize=(8, 6))
sns.jointplot(x='sepal_length', y='sepal_width', data=df, kind='scatter', color='blue')
plt.suptitle('Joint Plot with Customizations', fontsize=16)
plt.subplots_adjust(top=0.95)  # Adjust title to fit within the plot area
plt.savefig('joint_plot_customized.png', dpi=300)
plt.show()


#### Saving plot as png or svg
if we want to save any of these plots as images (e.g., PNG or SVG), 
we can simply use plt.savefig() right before plt.show()

In [61]:
plt.savefig('line_plot_customized.png', dpi=300)
plt.show()

<Figure size 640x480 with 0 Axes>

## 4- Real World Project 

#### importing libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np


#### importing file

In [ ]:
df = pd.read_csv("E:\\Programming\\internship\\week3assignment\\merch_sales.csv")
df.head()

#### Analyze the dataset
Summarize the dataset to understand its structure, missing values, and basic statistics.

In [ ]:
print("Dataset Summary:")
print(df.info())
print("\nBasic Statistics:")
print(df.describe())
print("\nMissing Values:")
print(df.isnull().sum())

#### Clean the Dataset:
**Handle missing values.**
**Remove duplicates or irrelevant data if necessary.**

In [ ]:
df['Review'] = df['Review'].fillna('No Review')  # Fill missing reviews with 'No Review'
# Remove duplicates
df = df.drop_duplicates()
# Perform exploratory data analysis (EDA)
print("\nEDA:")
print("Unique Product Categories:", df['Product Category'].unique())
print("Unique Buyer Genders:", df['Buyer Gender'].unique())
print("Unique Order Locations:", df['Order Location'].unique())

#### Analyze the Data:
Perform exploratory data analysis (EDA) to derive insights.
Generate advanced computations using NumPy.

In [ ]:
# Generate advanced computations using NumPy
average_sales_price = np.mean(df['Sales Price'])
print("\nAverage Sales Price:", average_sales_price)

#### Create Visualizations:
Generate insightful visualizations using Matplotlib.
Document findings and insights.

In [ ]:
# Create Visualizations
plt.figure(figsize=(14, 10))

# 1. Distribution of Sales Price
plt.subplot(2, 2, 1)
plt.hist(df['Sales Price'], bins=30, color='blue', edgecolor='black')
plt.title('Distribution of Sales Price')
plt.xlabel('Sales Price')
plt.ylabel('Frequency')

# 2. Sales by Product Category
plt.subplot(2, 2, 2)
df['Product Category'].value_counts().plot(kind='bar', color='green')
plt.title('Sales by Product Category')
plt.xlabel('Product Category')
plt.ylabel('Number of Sales')

# 3. Sales by Buyer Gender
plt.subplot(2, 2, 3)
df['Buyer Gender'].value_counts().plot(kind='bar', color='purple')
plt.title('Sales by Buyer Gender')
plt.xlabel('Buyer Gender')
plt.ylabel('Number of Sales')

# 4. Average Rating by Product Category
plt.subplot(2, 2, 4)
df.groupby('Product Category')['Rating'].mean().plot(kind='bar', color='orange')
plt.title('Average Rating by Product Category')
plt.xlabel('Product Category')
plt.ylabel('Average Rating')

plt.tight_layout()
plt.show()

#### Create a Dashboard:
Combine multiple visualizations into a dashboard using Matplotlib's subplots.

In [ ]:
# Create a Dashboard
fig, axs = plt.subplots(2, 2, figsize=(14, 10))

# 1. Distribution of Sales Price
axs[0, 0].hist(df['Sales Price'], bins=30, color='blue', edgecolor='black')
axs[0, 0].set_title('Distribution of Sales Price')
axs[0, 0].set_xlabel('Sales Price')
axs[0, 0].set_ylabel('Frequency')

# 2. Sales by Product Category
df['Product Category'].value_counts().plot(kind='bar', color='green', ax=axs[0, 1])
axs[0, 1].set_title('Sales by Product Category')
axs[0, 1].set_xlabel('Product Category')
axs[0, 1].set_ylabel('Number of Sales')

# 3. Sales by Buyer Gender
df['Buyer Gender'].value_counts().plot(kind='bar', color='purple', ax=axs[1, 0])
axs[1, 0].set_title('Sales by Buyer Gender')
axs[1, 0].set_xlabel('Buyer Gender')
axs[1, 0].set_ylabel('Number of Sales')

# 4. Average Rating by Product Category
df.groupby('Product Category')['Rating'].mean().plot(kind='bar', color='orange', ax=axs[1, 1])
axs[1, 1].set_title('Average Rating by Product Category')
axs[1, 1].set_xlabel('Product Category')
axs[1, 1].set_ylabel('Average Rating')

plt.tight_layout()
plt.show()

### 4.1 Documentation of the Above Real World Project ,for Sales Data Analysis Workflow with Each Steps Implementatio
---
#### 1. Define the Objective
Analyzed the sales data to identify key trends and insights such as high-performing product categories, customer demographics, and sales patterns.
---
#### 2. Collect and Prepare Data
- The dataset is already loaded into a DataFrame named `df`.
- **Columns and Data Types:**
Order ID int64 Order Date object Product ID object Product Category object Buyer Gender object Buyer Age int64 Order Location object International Shipping object Sales Price int64 Shipping Charges int64 Sales per Unit int64 Quantity int64 Total Sales int64 Rating int64 Review object
---
#### 3. Identify Data Types
- The dataset contains:
- **Categorical Variables:** `Product Category`, `Buyer Gender`, `Order Location`, `International Shipping`, `Review`.
- **Numerical Variables:** `Sales Price`, `Shipping Charges`, `Sales per Unit`, `Quantity`, `Total Sales`, `Rating`, `Buyer Age`.
---
#### 4. Choose Analysis Type
- **Univariate Analysis:** To explore distributions of individual variables like `Sales Price`, `Rating`, and `Buyer Gender`.
- **Bivariate Analysis:** To analyze relationships between variables, e.g., `Product Category` vs. `Sales Price` or `Rating` vs. `Total Sales`.
---
#### 5. Select Analysis Techniques
- **Descriptive Statistics:** Mean, median, standard deviation of numerical columns.
- **Visualizations:** Histograms, bar charts, and correlation heatmaps.
- **Correlation Analysis:** To measure relationships between numerical variables.
---
#### 6. Visualize the Data
`plt.figure(figsize=(14, 10))`
**Distribution of Sales Price**
`plt.subplot(2, 2, 1)
plt.hist(df['Sales Price'], bins=30, color='blue', edgecolor='black')
plt.title('Distribution of Sales Price')
plt.xlabel('Sales Price')
plt.ylabel('Frequency')`

**Sales by Product Category**
`plt.subplot(2, 2, 2)
df['Product Category'].value_counts().plot(kind='bar', color='green')
plt.title('Sales by Product Category')
plt.xlabel('Product Category')
plt.ylabel('Number of Sales')`

**Sales by Buyer Gender**
`plt.subplot(2, 2, 3)
df['Buyer Gender'].value_counts().plot(kind='bar', color='purple')
plt.title('Sales by Buyer Gender')
plt.xlabel('Buyer Gender')
plt.ylabel('Number of Sales')`

**Average Rating by Product Category**
`plt.subplot(2, 2, 4)
df.groupby('Product Category')['Rating'].mean().plot(kind='bar', color='orange')
plt.title('Average Rating by Product Category')
plt.xlabel('Product Category')
plt.ylabel('Average Rating')
plt.tight_layout()
plt.show()`

#### 7. Apply Analysis Methods
**Descriptive Statistics:** To summarize numerical columns (Sales Price, Total Sales, Rating).
**Correlation Analysis:** Calculating the correlation matrix for numerical variables.
`correlation_matrix = df.corr()
print("\nCorrelation Matrix:")
print(correlation_matrix)`

#### 8. Interpret the Results
**Distribution of Sales Price:** Understand the frequency of different price ranges.
**High-Selling Product Categories:** From the bar chart, identify which product categories have the most sales.
**Buyer Demographics:** Analyze the gender distribution to see the major contributing group.
**Correlation Insights:** Identify relationships, e.g., between Sales Price, Quantity, and Total Sales.

#### Implementation of Steps 4 to 8
**Univariate Analysis (Step 4):**

Histogram of Sales Price to visualize distribution.
Count plot for Product Category and Buyer Gender.
**Bivariate Analysis (Step 5):**
Bar chart to analyze the relationship between Product Category and average Rating.
**Analysis Techniques (Step 6):**
Use visualizations for patterns and correlations.
Generate a correlation matrix to assess numerical relationships.
**Insights and Results (Step 7 and 8):**
Identify categories with high average ratings and significant correlations, e.g., if higher Quantity leads to lower Sales Price per unit.

#### 9. Make Decisions
**Example decisions:**
Focus marketing efforts on high-selling product categories.
Target underrepresented buyer demographics for growth opportunities.
Optimize pricing strategies based on insights from the correlation matrix.

#### 10. Communicate Findings
Create a summary report with visualizations and key takeaways.
Share findings with stakeholders through charts and actionable insights.

#### 11. Take Action
Implement recommendations such as adjusting marketing budgets or improving customer service in top-performing categories.

#### 12. Refine and Iterate
Continuously monitor sales data and refine strategies based on updated insights and new patterns.

### Important Things to remember
**Data distribution:** describes the range of values in a dataset and how often each value occurs. It can be used to understand the data's center, spread, shape, and modality.
-***Data distribution can be represented graphically using histograms and box plots.***
-Two types of data distribution: discrete and continuous. Discrete data has specific values, while continuous data can have an infinite number of values.
**Data manipulation** is the process of changing or organizing data to make it more accessible, readable, and useful: 
**Purpose**
Data manipulation is a key step in data analysis and processing. It helps prepare data for analysis, reporting, visualization, and other computational tasks. 
**Techniques**
Data manipulation can involve a variety of operations, including filtering, sorting, aggregating, merging, and transforming data


### Missing data can be categorised into three types:
- **Missing Completely at Random (MCAR):** This means that the missing data
is not in any way related to any other data in the dataset.
- **Missing at Random (MAR):** This is when the missing data is somehow
related to other data in the dataset. 
- **Missing Not at Random (MNAR):** This occurs when there is a direct relation
to some missing data that the researchers haven’t measured.
**Standardization** is useful when you want to center your data with a mean of 0 and a standard deviation of 1, especially when your algorithm assumes that the data is normally distributed (e.g., linear regression, PCA).
## Methods to remove the Missing Data:
- **distribution of data**: statistics using mean,max,etc
- **imputation of KNN**:
- **mputation with linear regression**
This imputation technique utilises variables from the observed data to replace the
missing values with predicted values from a regression model.

## FEATURE SCALING: NORMALISATION AND STANDARDISATION
Another common problem we encounter when trying to analyse data is having
different units of measurement for a particular variable.
### Methods for scaling
- **Normalization**
- **Standard Deviation**

#### Example of Standardization
#### Original Data:
[10,12,14,16,18]
**Mean (𝜇):** 10+12+14+16+1/5=14 <br>

**Standard Deviation (𝜎):** Let's assume σ=2.83 for simplicity.<br>
**Standardization:** <br>
Using the formula <br>
`𝑋standardized=𝑋−𝜇/𝜎`<br>
For X=10:
𝑋standardized=10−14/2.83 =>−1.41<br>
for X=12:
Xstandardized=12−14/2.83=>−2<br>
Similar for all others aply same formula 
#### Standardized Data:
[-1.41, -0.71, 0, 0.71, 1.41]

Now, the data has a **mean of 0** and a **standard deviation of 1**.

**Normalization** is useful when you need to scale data into a specific range (usually [0, 1]) and works well for distance-based algorithms and neural networks.<br>
Formula:
`Xnormalized= X−Xmin/Xmax−Xmin` <br>
​Where:
X is the original value,
𝑋min
  is the minimum value in the dataset,
𝑋max​
  is the maximum value in the dataset.

`When to Use:`
`When features have different units or scales` (e.g., height in cm vs. weight in kg).
`Commonly used when applying algorithms that rely on distance metrics`, such as K-Nearest Neighbors (KNN) or Support Vector Machines (SVM).
`In neural networks where the activation functions` (like Sigmoid) expect values between 0 and 1. <br>
**But how do we know when**
`to use which one? Quite simply, you must standardise the data when it follows a Gaussian distribution. If not, normalise the data.`

Normalization and Standardization (Standard Deviation) are two common techniques used in data preprocessing to make datasets more comparable and improve machine learning model performance

## EDA (Explore Data Analysis)
When you're a beginner and find yourself stuck during the Exploratory Data Analysis (EDA) of a dataset, it's completely normal. EDA is a crucial step in data science that involves summarizing, visualizing, and interpreting the data to uncover patterns, anomalies, and insights. Here are some steps and tips to guide you through the process:

1. **Understand the Dataset**:
   - **Read the Data**: Use tools like pandas in Python to load the dataset (`pd.read_csv()`, `pd.read_excel()`, etc.).
   - **Check the Structure**: Use `df.info()` in pandas to get an overview of the data types, missing values, and the number of entries for each column.
   - **View the Data**: Use `df.head()` and `df.tail()` to see the first and last few rows of the dataset.

2. **Clean the Data**:
   - **Handle Missing Values**: Decide whether to fill them in, drop them, or use imputation techniques.
   - **Remove Duplicates**: Use `df.drop_duplicates()` to ensure you're working with unique data points.
   - **Correct Data Types**: Convert columns to the appropriate data types using `df['column'].astype()`.

3. **Summarize the Data**:
   - **Descriptive Statistics**: Use `df.describe()` to get summary statistics for numerical columns.
   - **Categorical Data Summary**: Use `df['category_column'].value_counts()` to see the distribution of categorical variables.

4. **Visualize the Data**:
   - **Histograms**: Plot histograms for numerical columns to understand their distribution.
   - **Box Plots**: Use box plots to identify outliers and the spread of the data.
   - **Bar Charts/Pie Charts**: Visualize the frequency or proportion of categorical data.
   - **Scatter Plots**: Plot scatter plots to explore relationships between two numerical variables.

5. **Ask Questions and Formulate Hypotheses**:
   - Think about what questions you want to answer with the data.
   - Formulate hypotheses based on your initial observations and domain knowledge.

6. **Dig Deeper**:
   - **Correlation Analysis**: Use `df.corr()` to see how numerical variables are related to each other.
   - **Grouping and Aggregation**: Use `groupby()` and aggregation functions to summarize data by categories.
   - **Feature Engineering**: Create new features from existing ones if they might be useful for your analysis.

7. **Document Your Findings**:
   - Keep a notebook or document where you record your observations, hypotheses, and the steps you took during EDA.
   - This will help you stay organized and communicate your findings effectively.

8. **Iterate**:
   - EDA is often an iterative process. As you uncover new insights, you may need to go back and refine your approach or explore new angles.

9. **Seek Help**:
   - Don't hesitate to ask for help from more experienced colleagues, online forums, or tutorials.
   - The data science community is generally very helpful and supportive.

Remember, the goal of EDA is to gain a deep understanding of your data so that you can make informed decisions about how to proceed with your analysis or modeling. Take your time, be methodical, and don't be afraid to explore different approaches.